Control/Monad/Logic/Class.hs

```diff
@@ -165,26 +181,110 @@ class (Monad m, Alternative m) => MonadLogic m where
     (>>-)      :: m a -> (a -> m b) -> m b
     infixl 1 >>-
 
-    -- | Logical __conditional.__ The equivalent of
-    --   <http://lpn.swi-prolog.org/lpnpage.php?pagetype=html&pageid=lpn-htmlse44 Prolog's soft-cut>.
-    --   If its first argument succeeds at all, then the results will be fed into
-    --   the success branch. Otherwise, the failure branch is taken.
-    --   satisfies the following laws:
-    --
-    --   > ifte (pure a) th el       == th a
-    --   > ifte empty th el          == el
-    --   > ifte (pure a <|> m) th el == th a <|> (m >>= th)
-    ifte       :: m a -> (a -> m b) -> m b -> m b
-
     -- | __Pruning.__ Selects one result out of many. Useful for when multiple
     --   results of a computation will be equivalent, or should be treated as
     --   such.
+    --
+    --   As an example, here's a way to determine if a number is
+    --   composite (has non-trivial integer division, i.e. the not a
+    --   prime number, see https://wikipedia.org/wiki/Composite_number):
+    --
+    --   > choose = foldr ((<|>) . pure) empty
+    --   >
+    --   > factors n = do a <- choose [2..n-1]
+    --   >                b <- choose [2..n-1]
+    --   >                guard (a * b == n)
+    --   >                pure (a, b)
+    --   >
+    --   > composite_ v = do _ <- factors v
+    --   >                   pure "Composite"
+    --
+    --   While this works as intended, it actually does too much work:
+    --
+    --   >>> observeAll (composite_ 20)
+    --   ["Composite", "Composite", "Composite", "Composite"]
+    --
+    --   Because there are multiple factors of 20, and they can also
+    --   occur in either order:
+    --
+    --   >>> observeAll (factors 20)
+    --   [(2,10), (4,5), (5,4), (10,2)]
+    --
+    --   Clearly one could just use 'observe' here to get the first
+    --   non-prime result, but if the call to @composite@ is in the
+    --   middle of other logic code then use 'once' instead.
+    --
+    --   > composite v = do _ <- once (factors v)
+    --   >                  pure "Composite"
+    --
+    --   >>> observeAll (composite 20)
+    --   ["Composite"]
+    --
     once       :: m a -> m a
 
     -- | __Inverts__ a logic computation. If @m@ succeeds with at least one value,
     --   @lnot m@ fails. If @m@ fails, then @lnot m@ succeeds with the value @()@.
+    --
+    --   For example, evaluating if a number is prime can be based on
+    --   the failure to find factors of a number:
+    --
+    --   > choose = foldr ((<|>) . pure) empty
+    --   >
+    --   > factors n = do a <- choose [2..n-1]
```

Comment:
Since we repeat this snippet thrice, could we simplify it a bit? Like
```haskell
divisors n = do d <- choose [2..n-1]
                guard (n `rem` d == 0)
                pure d 
```

Reply:
I left one of them (the first) in the longer form because it does show multiple backtracking points, and I think it's good as a demonstrator that there can be multiple obvious logic backtracking points in a computation.